test/test_ops.py

```diff
@@ -1450,5 +1451,140 @@ def test_is_leaf_node(self, dim, p, block_size, inplace):
         assert len(graph_node_names[0]) == 1 + op_obj.n_inputs
 
 
+def focal_loss_device_and_dtype():
+    # This is a helper function to provide list of device and dtype pair for TestFocalLoss
+    # We need this because torch.half is not fully supported on cpu
+    # For instance, torch.log(torch.rand(1, dtype=torch.half)) will produce error currently
+    def augment_need_cuda_param(x):
+        if x[0] == "cuda":
+            return pytest.param(x, marks=pytest.mark.needs_cuda)
+        return x
+
+    result = [
+        (device, dtype)
+        for device in ["cpu", "cuda"]
+        for dtype in [torch.float32, torch.half]
+        if (device, dtype) != ("cpu", torch.half)
+    ]
+    result = [augment_need_cuda_param(x) for x in result]
+    return result
```

Comment:
Not sure about this. This adds a lot more complexity than

```py
@cpu_and_gpu
@pytest.mark.parametrize("dtype", [torch.float32, torch.half])
def test_foo(device, dtype):
    if device.type.startswith("cuda"):
        return
    ...
```

Was there a reason to not go this way?